src/lib/logical_query_plan/projection_node.cpp

```diff
@@ -13,7 +13,7 @@
 
 namespace opossum {
 
-ProjectionNode::ProjectionNode(const std::vector<std::shared_ptr<Expression>>& column_expressions)
+ProjectionNode::ProjectionNode(std::vector<std::shared_ptr<Expression>>&& column_expressions)
```

Comment:
This seems to be a loooong discussion: https://stackoverflow.com/questions/37935393/pass-by-value-vs-pass-by-rvalue-reference

I kind of like this answer: https://stackoverflow.com/a/37953536/2204581

Since we are passing in pointers, not actual expressions, the copy of that vector is essentially meaningless. Both caller and callee are still working on the same expressions. Because of that, it makes sense to me that the caller has to give up his ownership of the expressions.

Reply:
> the vector would allow the caller to modify it.

Not exactly, just his own copy.

Reply:
>> the vector would allow the caller to modify it.
> Not exactly, just his own copy.
His own copy *of shared_ptrs* that point to the objects that are now managed by the ProjectionNode.

This is not only about the cost of copying, it is also about giving up the shared pointers. In some places, the copy semantics of `shared_ptr`s are convenient, but semantically, I am not a fan. They don't really copy anything that we care about.

Reply:
Okay but whatever is passed should be a vector of unique pointers. Are expressions trees or graphs?

Reply:
Trees.

We cannot create expressions as `unique_ptr` because they are inheriting from `enable_shared_from_this`.

Reply:
Yes and they inherit from `enable_shared_from_this` because of two lines in `set_left_child` and `set_right_child` where we set the parent. In a tree, the parent could easily be the owner of its children. `_left_child` and `_right_child` could be `unique_ptr`s and `_parent` a normal non-owning raw-pointer. `std::weap_ptr` wouldn’t be needed, because a child would be certain that as long as it exists itself, its parent must surely also exist. Then the root of the expression tree could be stored in a `unique_ptr` too and the semantics would be clearer.

Reply:
Okay at the current state of the conversation, can we agree that we pass by value as a default irrespective of is stored in the vector because the lifetime of the vector is separate from the lifetime of the objects that the pointers reference?

Reply:
We had the raw pointer discussion in #398 and this does not really change things. Unless something changes fundamentally, we will not have any raw pointers.

Reply:
In this case, passing it by value does not hurt too much. How would we adapt the flowchart? We don't want people to pass `ValueColumn<int>::values()` by copy.